controllers/agents/app/agentapp_controller.go

```diff
@@ -0,0 +1,136 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// Agent Application Reconciler reconciles a Agent Application object
+type AgentApplicationReconciler struct {
+	client.Client
+}
+
+func NewAgentApplicationReconciler(mgr ctrl.Manager) *AgentApplicationReconciler {
+	return &AgentApplicationReconciler{
+		Client: mgr.GetClient(),
+	}
+}
+
+const agentappfinalizer = "agentapp.primaza.io/finalizer"
+
+func (r *AgentApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+	l.Info("Reconcile Agent Service Deployment")
+
+	// first, get the agent app deployment
+	agentappdeployment := appsv1.Deployment{}
+	objKey := client.ObjectKey{
+		Name:      constants.ApplicationAgentDeploymentName,
+		Namespace: req.Namespace,
+	}
+	err := r.Get(ctx, objKey, &agentappdeployment)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// it should never happen that this controller does not find itself
+			l.Error(err,
+				"application agent deployment not found, that should be a bug",
+				"expected deployment name", constants.ApplicationAgentDeploymentName)
+			panic(err)
+		}
+		return ctrl.Result{}, err
+	}
+	// TODO: We need to ensure that we do not reconcile service catalog and service binding once the agent application deployment
+	// is marked for deletion.
+	if !agentappdeployment.DeletionTimestamp.IsZero() {
+		if err := r.removePrimazaResources(ctx, req); err != nil {
+			l.Error(err, "Service catalog and service binding deletion failed")
+			return ctrl.Result{}, err
+		}
+		// remove the finalizer so we don't requeue
+		if controllerutil.RemoveFinalizer(&agentappdeployment, agentappfinalizer) {
+			if err = r.Update(ctx, &agentappdeployment, &client.UpdateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else if !controllerutil.ContainsFinalizer(&agentappdeployment, agentappfinalizer) {
+		// add a finalizer since we have deletion logic
+		if controllerutil.AddFinalizer(&agentappdeployment, agentappfinalizer) {
+			if err = r.Update(ctx, &agentappdeployment, &client.UpdateOptions{}); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+func (r *AgentApplicationReconciler) removePrimazaResources(ctx context.Context, req ctrl.Request) error {
+	errs := []error{}
+	if err := r.removeServiceCatalog(ctx, req); err != nil {
+		errs = append(errs, err)
+	}
+	if err := r.removeServiceBinding(ctx, req); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+func (r *AgentApplicationReconciler) removeServiceCatalog(ctx context.Context, req ctrl.Request) error {
+	// first, get the service catalog
+	servicecatalogList := v1alpha1.ServiceCatalogList{}
+	if err := r.List(ctx, &servicecatalogList, &client.ListOptions{Namespace: req.Namespace}); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	var errorList []error
+	for _, scat := range servicecatalogList.Items {
```

Comment:
Instead of looping through the list using key and value, you can get the index of ervicecatalogList.Items with a single variable. and then initialize just the service catalog object inside the loop.
for example:
`or index := range clusterEnvironmentList.Items {
		ce := clusterEnvironmentList.Items[index]`